Feature/warehouses

GianSz-meli opened the pull request:
In these commits I refactored the application in order to use mysql db to store the data (warehouses). Also, I added a validation to check wether a locality exists or not based on the id given on the creation request.

GianSz-meli committed:
Refactor: Changed contracts and implementations to work with err

GianSz-meli committed:
Feat: Added sql repository implementation

GianSz-meli committed:
Feat: Added validation of locality existence

Added a validation to check if a locality exists before the warehouse is sent to the database

estfernandez-meli commented on pkg/errors/errors.go:
```diff
@@ -65,14 +67,37 @@ func getMappedError(err error) *ApiError {
 	}
+
+	var mysqlErr *mysql.MySQLError
+	if errors.As(err, &mysqlErr) {
```
Aun no se define como manejar los errores de mysql

estfernandez-meli commented on internal/repository/locality/locality_mysql.go:
```diff
@@ -0,0 +1,55 @@
+	)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, nil
```
No se retorna el error

estfernandez-meli commented on internal/handler/warehouse_handler.go:
```diff
@@ -25,7 +25,11 @@ func NewWarehouseHandler(warehouseService warehouse.WarehouseService) *Warehouse
 func (h *WarehouseHandler) GetAllWarehouses(w http.ResponseWriter, r *http.Request) {
-	warehouses := h.warehouseService.GetAllWarehouses()
+	warehouses, err := h.warehouseService.GetAllWarehouses()
+	if err != nil {
```
Como sugerencia, hacer log.Println del error para que no perdamos contexto del error de mysql

estfernandez-meli commented on pkg/models/locality.go:
```diff
@@ -0,0 +1,21 @@
+package models
+
+// Country model according to database schema
+type Country struct {
```
Campos de Country, Province y locality no pueden ser punteros

estfernandez-meli commented on internal/repository/locality/locality_mysql.go:
```diff
@@ -0,0 +1,55 @@
+		return nil, err
+	}
+
+	province.Country = &country
```
Si el country es nil, no deberia existir province, y viceversa

GianSz-meli committed:
Merge branch 'develop' into feature/warehouses

GianSz-meli committed:
Refactor: Changed locality model and refactored repositories

Data type of locality, province and country were changed since they shouldn't be pointers as they are not optional. Also, grouped warehouse db queries on a single file and modified repository logic to work with the new model. Lastly, modified getById to return the error gotten from the repo

estfernandez-meli commented on internal/repository/warehouse/warehouse_mysql.go:
```diff
@@ -0,0 +1,189 @@
+}
+
+func (r *SqlWarehouseRepository) GetAll() ([]models.Warehouse, error) {
+	query := `
```
Las querys deberian ser constantes, y ojalá en un archivo diferente

estfernandez-meli commented on pkg/models/warehouse.go:
```diff
@@ -8,24 +8,45 @@ type Warehouse struct {
+		MinimumTemperature: w.MinimumTemperature,
+		Locality:           w.Locality,
+	}
+	if w.Locality != nil {
```
Locality tiene LocalityName y Province. Así como está solo agrega el id a la struct

GianSz-meli committed:
Merge branch 'develop' into feature/warehouses

GianSz-meli committed:
Refactor: Deleted locality repo and moved validation to warehouse repo

Deleted locality repository since there is not need to use loclaity repository to check data validity manually, now the repository is in charged of this by catching and handling sql errors that are later passed through the service and handler

GianSz-meli committed:
Refactor: Moved data validation on deletion from service to repo
